pkg/registry/pod/rest.go

```diff
@@ -81,9 +81,28 @@ func (podStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fiel
 	return append(errorList, validation.ValidatePodUpdate(obj.(*api.Pod), old.(*api.Pod))...)
 }
 
-// CheckGracefulDelete allows a pod to be gracefully deleted.
+// CheckGracefulDelete allows a pod to be gracefully deleted. It updates the DeleteOptions to
+// reflect the desired grace value.
 func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
-	return false
+	if options == nil {
+		return false
+	}
+	pod := obj.(*api.Pod)
+	// use the user provided value
+	if options.GracePeriodSeconds != nil {
+		period := int64(*options.GracePeriodSeconds)
+		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
+		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
+			pod.Spec.TerminationGracePeriodSeconds = &period
```

Comment:
> On May 27, 2015, at 4:51 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/registry/pod/rest.go:
> 
> >  func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
> > -	return false
> > +	if options == nil {
> > +		return false
> > +	}
> > +	pod := obj.(*api.Pod)
> > +	// use the user provided value
> > +	if options.GracePeriodSeconds != nil {
> > +		period := int64(*options.GracePeriodSeconds)
> > +		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
> > +		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
> > +			pod.Spec.TerminationGracePeriodSeconds = &period
> The main problem I have with changing TerminationGracePeriodSeconds is that it would generate a conflict with user-provided configuration, in violation of the API rules:
> https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conventions.md#late-initialization
> 
Isn't he grace period provided when the user deletes an explicit modification of the user's desire for the pod?  The user is saying "I want this pod to no longer exist in 10 seconds from now", which is very close to what TerminationGracePeriodSeconds seems to mean.  Preserving the original value seems to have no value, because delete from a client is user intent.

I'm not hardline on this but since delete is user intent in many cases it seems to overlap the purpose of the spec field.  A user might update the spec field, then call delete - the delete options area convenient way to avoid having to make two requests.
> Also, where are we going to record GracePeriodSeconds for other object kinds?
> 
In something in spec that makes sense for them?  The rules for graceful termination (and description) is likely to vary quite significantly - for instance, I could imagine that a job termination has a very different meaning than the termination of the pods under it.  Not every resource has to support graceful termination of course.
> I view TerminationGracePeriodSeconds as the user-specified default for use by controllers, when the user can't directly provide a more specific GracePeriodSeconds upon deletion.
> 
Or the default the user will use - the author of a pod spec != the person who creates and deletes it.
> If we need another field for recording the (relative) GracePeriodSeconds, we should add one. We could keep DeletionTimestamp, or replace it. I imagine we'll also want an LastUpdatedTimestamp, which might be a sufficient replacement for debugging purposes (not that we have to implement that now).
> 
But last updated won't be sufficient to track the delta because updates can still be made after deletion - we'd need a specific timestamp anyway for the delta to apply to (the time the item was requested to be deleted).
> —
> Reply to this email directly or view it on GitHub.
>